analytics-ios-mcvid/Classes/SEGMCVIDTracker.m

```diff
@@ -32,10 +37,169 @@ -(id)initWithOrganizationId:(NSString *)organizationId region:(NSString *)region
       self.organizationId = organizationId;
       self.region = region;
     }
+
+    //Values for exponential backoff retry logic for API calls
+    NSUInteger maxRetryCount = 11;
+    NSUInteger currentRetryCount = 1;
+    NSUInteger maxRetryTimeSecs = 300;
+    self.backgroundQueue = dispatch_queue_create("com.segment.mcvid", NULL);
+
+
+    //Store advertisingId and marketingCloudId on local storage
+    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
+    NSString *cachedMarketingCloudId = [defaults stringForKey:@"MarketingCloudId"];
+    NSString *cachedAdvertisingId = [defaults stringForKey:@"AdvertisingId"];
+    [[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString];
+    //This is was SEGIDFA() is going under the hood. NSString *idfaString = [[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString];
+    NSString *segIdfa = SEGIDFA();
+    if (![cachedAdvertisingId isEqualToString:segIdfa]) {
+        [defaults setObject:segIdfa forKey:@"AdvertisingId"];
+    }
+
+    if (cachedMarketingCloudId.length == 0) {
+        [self getMarketingCloudId:organizationId maxRetryCount:maxRetryCount currentRetryCount:currentRetryCount maxRetryTimeSecs:maxRetryTimeSecs completion:^(NSString *marketingCloudId, NSError *error) {
+          [defaults setObject:marketingCloudId forKey:@"MarketingCloudId"];
+          [self syncMarketingCloudId:cachedAdvertisingId organizationId:organizationId marketingCloudId:cachedMarketingCloudId maxRetryCount:maxRetryCount currentRetryCount:currentRetryCount maxRetryTimeSecs:maxRetryTimeSecs completion:^(NSError *error) {
```

Comment:
🚨 There's another bug here: `cachedMarketingCloudId` will always be `nil` at this point. Due to this bug and the one a few lines above both the marketing cloud id and the advertising id will be `nil`, thus making the id-sync request invalid.